import { Request, Response, NextFunction } from 'express';
import { User } from '../models';
import { AppError } from '../middleware';
import { ResponseMapper } from '../utils/response-mapper';

const responseMapper = ResponseMapper.getInstance();

export const createUser = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const user = await User.create(req.body);
    return responseMapper.created(res, responseMapper.mapUser(user));
  } catch (error) {
    next(error);
  }
};

export const getUsers = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const users = await User.find();
    return responseMapper.success(res, responseMapper.mapUsers(users));
  } catch (error) {
    next(error);
  }
};

export const getUser = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const user = await User.findById(req.params.id);
    if (!user) {
      return next(new AppError('No user found with that ID', 404));
    }
    return responseMapper.success(res, responseMapper.mapUser(user));
  } catch (error) {
    next(error);
  }
};

export const updateUser = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const user = await User.findByIdAndUpdate(req.params.id, req.body, {
      new: true,
      runValidators: true,
    });
    if (!user) {
      return next(new AppError('No user found with that ID', 404));
    }
    return responseMapper.success(res, responseMapper.mapUser(user));
  } catch (error) {
    next(error);
  }
};

export const deleteUser = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const user = await User.findByIdAndDelete(req.params.id);
    if (!user) {
      return next(new AppError('No user found with that ID', 404));
    }
    return responseMapper.noContent(res);
  } catch (error) {
    next(error);
  }
}; 